### PR TITLE
[native_assets_cli] `BuildConfig.linkingEnabled`

### PR DESCRIPTION
**Pull request by dcharkes**
Step 1 of

* https://github.com/dart-lang/native/issues/1252

Naming bike shedding:

* `BuildConfig.hasLinkPhase`?
* `BuildConfig.willLink`?
* `BuildConfig.isLinkHookEnabled`?

### CODE REVIEW

File: pkgs/native_assets_cli/test/model/checksum_test.dart [hunk 1]
```diff
@@ -76,9 +83,26 @@ void main() {
...
+        targetOS: OSImpl.linux,
+        buildMode: BuildModeImpl.release,
+        linkModePreference: LinkModePreferenceImpl.dynamic,
+        supportedAssetTypes: [NativeCodeAsset.type],
+        hook: Hook.build,
+        version: HookConfigImpl.latestVersion,
+        hasLinkPhase: false,
+      );
+      printOnFailure([name1, name5].toString());
+      expect(name1 != name5, true);
```
Comment (HosseinYousefi) [inline pkgs/native_assets_cli/test/model/checksum_test.dart:105]: nit: `notEquals`

-- Could be find-&-replaced in another PR.

**Comment by mosuem** [comment]
> Step 1 of
> 
> * [[native_assets_cli] Stop running link hooks in JIT mode. #1252](https://github.com/dart-lang/native/issues/1252)
> 
> Naming bike shedding:
> 
> * `BuildConfig.hasLinkPhase`?
> * `BuildConfig.willLink`?
> * `BuildConfig.isLinkHookEnabled`?

IMO, `phase` sounds too technical for no reason, `willLink` could be misleading (what if there is no link hook?), so `isLinkEnabled` or similar should be fine.

**Comment by HosseinYousefi** [comment]
> what if there is no link hook?

Good point, but the same thing could be said about the other names. The only way to mitigate this confusion completely would be to use verbs that show possibility instead of down right stating that things will be linked. For example `supportsLinkHook` or `canBeLinked`.

**Comment by mosuem** [comment]
> the same thing could be said about the other names. The only way to mitigate this confusion completely would be to use verbs that show possibility instead of down right stating that things will be linked. 

I agree, I think my proposed `isLinkEnabled` also works in this regard. It does not promise that linking will happen, only that it is enabled in the config.

**Comment by dcharkes** [comment]
> > the same thing could be said about the other names. The only way to mitigate this confusion completely would be to use verbs that show possibility instead of down right stating that things will be linked.
> 
> I agree, I think my proposed `isLinkEnabled` also works in this regard. It does not promise that linking will happen, only that it is enabled in the config.

Much better! Bouncing off this with some other wording that feels slightly more natural:

* `linkingEnabled`
* `linkHooksEnabled`

I think, I lean towards the first.

**Comment by mosuem** [comment]
For me, `support`/`allow`/`can` sounds more like a general option, less like a switch.

+1 for `linkingEnabled`

**Comment by dcharkes** [comment]
Hooks are called/invoked/run or not called/invoked/run.

I think the "enabled" is whether they will be called/invoked/run.

"allow" is for access control, "enable" is for possibility. So, I think `linkingEnabled` is better than `linkingAllowed`.

> `canLink`

I like this, together with `addAsset(..., {bool linkInPackage})`. "can" and "enabled" are the same thing to me, but can definitely omits the possibility that something has already happened. And it's concise. 👌

**Comment by HosseinYousefi** [comment]
> For me, `support`/`allow`/`can` sounds more like a general option, less like a switch.
> 
> +1 for `linkingEnabled`

I don't want to keep bike-shedding, so feel free to go with that name. However "enabled" suggests that there already is a linking step involved, so I "must", build linkable things.

Something like `linkingAvailable` suggests that I have the possibility of building linkable things.

**Comment by dcharkes** [comment]
> `linkingAvailable`

I also like this one!

"available" and "enabled" is maybe better than "can", because can leaves it open if you as hook author can invoke linking or if the system might possibly be linking.

"available" is maybe better than "enabled" because it clearly states that you can use something from your hook. While enabled could mean something has already happened.

(Okay, I'm going to keep changing my mind in this conversation.)

@mosuem `linkingAvailable`?

File: pkgs/native_assets_builder/lib/src/build_runner/build_runner.dart [hunk 2]
```diff
@@ -201,6 +206,7 @@ class NativeAssetsBuildRunner {
     BuildModeImpl buildMode,
     LinkModePreferenceImpl linkModePreference,
     DependencyMetadata? dependencyMetadata,
+    bool? hasLinkPhase,
```
[the remark this answers is not in the document]
Comment (dcharkes) [reply pkgs/native_assets_builder/lib/src/build_runner/build_runner.dart:209]: Yeah, I prefer named args as well, but Moritz recently changed this to positional, so not going to change it back now.

File: pkgs/native_assets_cli/test/api/build_config_test.dart [hunk 3]
```diff
@@ -194,6 +203,34 @@ void main() async {
...
+      packageRoot: tempUri,
+      targetArchitecture: Architecture.x64,
+      targetOS: OS.windows,
+      buildMode: BuildMode.release,
+      linkModePreference: LinkModePreference.preferStatic,
+      hasLinkPhase: false,
+    );
+
+    expect(buildConfig1, equals(buildConfig1));
+    expect(buildConfig1.hashCode == buildConfig2.hashCode, false);
```
[the remark this answers is not in the document]
Comment (dcharkes) [reply pkgs/native_assets_cli/test/api/build_config_test.dart:230]: You mean `isNot`? There is no matcher `notEquals`.

[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply pkgs/native_assets_cli/test/api/build_config_test.dart]: Hehe, yes unfortunately github PR comments don't have auto-suggestions 🙈

**Comment by mosuem** [comment]
There is a distinction between the link.dart file and the linking step. The link.dart file is certainly always available, but not always enabled, and the linking step is not always available. So I would say that either `linkingStepAvailable`/`linkingPhaseAvailable` or `linkingEnabled` is fine by me.

I do think `linkingEnabled` is shorter and simpler, as we don't need to distinguish between link hook file and phase for `enabled`, but we do for `available`. Maybe I am overthinking this.

**Comment by dcharkes** [comment]
I was asking our trusty assistant earlier:


Option | Focus | Nuance | Clarity
-- | -- | -- | --
linkingAvailable | Capability & Presence | Suggests linking is an option, but not necessarily the default or always guaranteed. | Very clear
canLink | Capability & Potential | Emphasizes the possibility of linking, but other factors might influence the outcome. | Clear
linkingEnabled | Configuration & Intent | Indicates linking is explicitly allowed and expected due to the build configuration. | Very clear

The Dart and Flutter SDK have the capability to link in some configurations (the configurations in which Dart is build in AOT mode). So seeing from the point of view that build hooks are invoked with a variety of different `BuildConfig`s, "enabled" seems to be a more fitting word than "available".

> I don't want to keep bike-shedding, so feel free to go with that name. However "enabled" suggests that there already is a linking step involved, so I "must", build linkable things.

I will add a doc comment to clarify that one "may" build linkable things instead of "must".

Thanks for all the input!

**Comment by HosseinYousefi** [comment]
> There is a distinction between the link.dart file and the linking step. The link.dart file is certainly always available, but not always enabled, and the linking step is not always available. So I would say that either `linkingStepAvailable`/`linkingPhaseAvailable` or `linkingEnabled` is fine by me.

Good point, I didn't think of it that way. 

> I do think `linkingEnabled` is shorter and simpler, as we don't need to distinguish between link hook file and phase for `enabled`, but we do for `available`. Maybe I am overthinking this.

I think we all like `linkingEnabled` enough to go with that.

**Review by mosuem — APPROVED** [review]

**Pull request merged**